ds701_book/06-Clustering-I-kmeans.qmd

```diff
@@ -312,19 +321,173 @@ The other from the top 10 were SVM, Apriori, EM, PageRank, AdaBoost, kNN, Naive
 ## The $k$-means algorithm
 :::
 
-1. Pick $k$ cluster centers $\{c_1, \dots, c_k\}$.  
-    - These can be chosen randomly, or by some other method.
-2. For each $j$, define the cluster $X_j$ as the set of points in $X$ that are <font color="blue">closest to center</font> $c_j$.  
+1. Pick $K$ cluster centers $\{c_1, \dots, c_K\}$.  
+    - These can be randomly chosen datapoints, or by some other method such as
+     [$k$-means++](https://en.wikipedia.org/wiki/K-means%2B%2B)
+2. For each $j$, define the cluster $C_j$ as the set of points in $X$ that are <font color="blue">closest to center</font> $c_j$.  
     - Nearer to $c_j$ than to any other center.
-3. For each $j$, update $c_j$ to be <font color="blue">the center of mass of cluster</font> $X_j$.  
-    - In other words, $c_j$ is the mean of the vectors in $X_j$.
-4. Repeat (ie, go to Step 2) until convergence^[Convergence to be defined].
+3. For each $j$, update $c_j$ to be <font color="blue">the center of mass of cluster</font> $C_j$.  
+    - In other words, $c_j$ is the mean of the vectors in $C_j$.
+4. Repeat (ie, go to Step 2) until convergence
+    - either the cluster centers change below a threshold, or inertia changes below a threshold or a maximum number of iterations is reached.
 
 ---
 
-Let's see this in practice:
+Let's see this in practice with well separated clusters and also look at the WCSS:
+
+```{python}
```

Comment:
You can put these lines under ```{python}

```
#| fig-align: center
#| fig-width: 75%
#| fig-wdith: 75%
```

More details on figs here 

- https://quarto.org/docs/authoring/figures.html#computations
- https://quarto.org/docs/computations/execution-options.html#figure-options

Reply:
didn't work. can we merge and fix it later?

Reply:
Yes, let's leave it to fix later. Just finished my review.

Reply:
ok